### PR TITLE
Ex1

### PR DESCRIPTION
**Pull request by danieTalmon**

### CODE REVIEW

File: src/app/product/product-list/product-list.component.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+export class ProductListComponent implements OnInit {
+
+  products: Product[];
+  isProductInCart: Record<string, boolean>;
+
+
+  constructor(private productService: ProductService, private cartService: CartService) { }
+
+  ngOnInit() {
+     this.productService.getProducts$().subscribe((products) => {
```
Comment (alon6699) [inline src/app/product/product-list/product-list.component.ts:30]: use combineLatest convert to {product: Product, isInCart: boolean}[]

File: src/app/product/product-list/product-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import { CartService } from './../../services/cart/cart.service';
+import { ProductService } from 'src/app/services/product/product.service';
+import { Component, OnInit } from '@angular/core';
+import { BehaviorSubject, Observable, of } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+export interface Product {
```
Comment (alon6699) [inline src/app/product/product-list/product-list.component.ts:7]: extract to file ... put inside models folder , product.model.ts

File: src/app/cart/cart.component.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      this.productService.getProduct(productName).subscribe(product => {
+        fullDetailsCart.push(product);
+      });
+    });
+    this.fullDetailsCart = fullDetailsCart;
+    }
+
+  removeFromCart(productName: string) {
+    this.cartService.removeFromCart(productName);
+    this.isInCart[productName] = false;
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:50]: remove

File: src/app/cart/cart.component.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    this.productService.updateIsInCart(this.isInCart);
+  }
+
+  changeAmount(productAmount: ProductAmount) {
+    this.cart[productAmount.name] = productAmount.amount;
+    this.cartService.updateCart(this.cart);
+  }
+
+  totalPrice() {
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:59]: type - observable

File: src/app/cart/cart.component.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    return this.fullDetailsCart.reduce((acc: number, product: Product) => {
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:60]: extract to variable

File: src/app/cart/cart.component.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      if (product) {
+        return acc + this.cart[product.name] * product.price;
+      }
+    }, 0);
+  }
+
+  checkout() {
+    this.products.forEach(product => {
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:68]: move this logic inside product service

File: src/app/cart/cart.component.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      if (product.limit) {
+        product.limit -= this.cart[product.name];
+      }
+    });
+    this.productService.updateProducts(this.products);
+
+    Object.keys(this.cart).forEach(productName => {
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:75]: use checkout method from ts execrise

File: src/app/product/product-info/product-info.component.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
+<div>
+  <div id="item-grid" class="ui one column center aligned middle aligned page grid">
+    <div class="column twelve wide">
+      <div class="ui centered aligned items">
+        <div class="ui centered aliened item" *ngIf="product">
+            <img class= "ui medium image" src="../../assets/images/{{product.name}}.jpg">
```
Comment (alon6699) [inline src/app/product/product-info/product-info.component.html:6]: [src]

File: src/app/product/product-info/product-info.component.html [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            <div id="product-header" class="ui large header"> {{product.name}}</div>
+            <div class="description">
+              <div class="ui grid">
+                <div class="row">
+                  <div class="ui medium header"> {{product.description}} </div>
+                </div>
+                <div class="row">
+                  <div class="ui four wide column">
+                    <div class="ui fluid small header">
+                      <i class="small shekel sign icon compact-icon"></i>{{product.price}}
```
Comment (alon6699) [inline src/app/product/product-info/product-info.component.html:17]: use currecny pipe

File: src/app/product/product-info/product-info.component.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+})
+export class ProductInfoComponent implements OnInit {
+  product: Product;
+
+  constructor(private productService: ProductService, private route: ActivatedRoute) { }
+
+  ngOnInit() {
+    const productName: string = this.route.snapshot.paramMap.get('id');
+    this.productService.getProduct(productName).subscribe((product) => {
+      this.product = product;
```
Comment (alon6699) [inline src/app/product/product-info/product-info.component.ts:19]: use async pipe

File: src/app/product/product-info/product-info.component.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  templateUrl: './product-info.component.html',
+  styleUrls: ['./product-info.component.less']
+})
+export class ProductInfoComponent implements OnInit {
+  product: Product;
+
+  constructor(private productService: ProductService, private route: ActivatedRoute) { }
+
+  ngOnInit() {
+    const productName: string = this.route.snapshot.paramMap.get('id');
```
Comment (alon6699) [inline src/app/product/product-info/product-info.component.ts:17]: without snapshot .. use param observable

File: src/app/product/product-list/product-list.component.html [hunk 12]
```diff
@@ -0,0 +1,4 @@
+<div *ngIf="products && isProductInCart"  class="ui link cards four column stretched grid padded">
```
Comment (alon6699) [inline src/app/product/product-list/product-list.component.html:1]: remove ngIf

File: src/app/product/product-list/product-list.component.html [hunk 13]
```diff
@@ -0,0 +1,4 @@
...
+  <ar-product *ngFor="let product of products" [product]="product" [isInCart]="isProductInCart[product.name]" (add)="addToCart($event)" (remove)="removeFromCart($event)" class="column" >
```
Comment (alon6699) [inline src/app/product/product-list/product-list.component.html:2]: use aync pipe

Comment (alon6699) [inline src/app/product/product-list/product-list.component.html:2]: $event -> product.name

File: src/app/product/product-list/product-list.component.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+  price: number;
+  amount: number;
+  limit?: number;
+}
+
+
+@Component({
+  selector: 'ar-product-list',
+  templateUrl: './product-list.component.html',
+  styleUrls: ['./product-list.component.less']
```
Comment (alon6699) [inline src/app/product/product-list/product-list.component.ts:19]: remove

File: src/app/product/product.component.html [hunk 15]
```diff
@@ -0,0 +1,40 @@
+<div class="ui fluid card" *ngIf="product" >
+    <div class="image">
+        <img class="card-image" src="../../assets/images/{{product.name}}.jpg">
```
Comment (alon6699) [inline src/app/product/product.component.html:3]: [src]

File: src/app/product/product.component.html [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+          <span class="left floated">
+          <i class="shekel sign icon"></i>{{product.price}}
+          </span>
+        </div>
+        <div class="item">
+          <span class="left floated">
+            <p>3/5</p>
+          </span>
+          <span class="right floated">
+            <button (click)='isInCart? removeFromCart(product.name): addToCart(product.name)'
```
Comment (alon6699) [inline src/app/product/product.component.html:32]: do not send product.name

Comment (alon6699) [inline src/app/product/product.component.html:32]: use two buttons

File: src/app/product/product.component.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+
+  @Input()
+  product: Product;
+
+  @Input()
+  isInCart: boolean;
+
+  @Output()
+  add: EventEmitter<string> = new EventEmitter<string>();
```
Comment (alon6699) [inline src/app/product/product.component.ts:20]: change to void and put inside the constructor

File: src/app/services/cart/cart.service.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
+import { BehaviorSubject, Observable, of } from 'rxjs';
+import { Injectable } from '@angular/core';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CartService {
+
+  private cart$: BehaviorSubject<Record<string, number>> = new BehaviorSubject<Record<string, number>>({});
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:9]: constructor

File: src/app/services/cart/cart.service.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
+import { BehaviorSubject, Observable, of } from 'rxjs';
+import { Injectable } from '@angular/core';
+
+@Injectable({
+  providedIn: 'root'
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:5]: remove

File: src/app/services/cart/cart.service.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+})
+export class CartService {
+
+  private cart$: BehaviorSubject<Record<string, number>> = new BehaviorSubject<Record<string, number>>({});
+
+  constructor() { }
+
+  getCart(): Observable<Record<string, number>> {
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:13]: getCart$()

File: src/app/services/cart/cart.service.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  changeAmount(productName: string, newAmount: number, limit: number | null) {
+    let cart: Record<string, number> = this.cart$.getValue();
+    if (!limit || newAmount < limit) {
+      cart[productName] = newAmount;
+    }
+  }
+
+  updateCart(cart: Record<string, number>) {
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:39]: remove

File: src/app/services/cart/cart.service.ts [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+
+  removeFromCart(productName: string) {
+    let cart: Record<string, number> = this.cart$.getValue();
+    delete cart[productName];
+
+
+    this.updateCart(cart);
+  }
+
+  changeAmount(productName: string, newAmount: number, limit: number | null) {
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:32]: emit new cart in this method

File: src/app/services/cart/cart.service.ts [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+    let cart: Record<string, number> = this.cart$.getValue();
+    cart[productName] = 1;
+
+    this.updateCart(cart);
+  }
+
+  removeFromCart(productName: string) {
+    let cart: Record<string, number> = this.cart$.getValue();
+    delete cart[productName];
+
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:27]: remove spaces

File: src/app/services/cart/cart.service.ts [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    if (!limit || newAmount < limit) {
+      cart[productName] = newAmount;
+    }
+  }
+
+  updateCart(cart: Record<string, number>) {
+    this.cart$.next(cart);
+  }
+
+  totalProducts() {
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:43]: type, return observable

File: src/app/services/product/product.service.ts [hunk 25]
```diff
@@ -0,0 +1,58 @@
+import { map, catchError } from 'rxjs/operators';
+import { of, Observable, BehaviorSubject } from 'rxjs';
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import * as ProductJson from 'src/assets/products.json';
+import { Product } from 'src/app/product/product-list/product-list.component';
+
+
+@Injectable({
+  providedIn: 'root'
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:10]: remove

File: src/app/services/product/product.service.ts [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+})
+export class ProductService {
+  private products$: BehaviorSubject<Product[]> = new BehaviorSubject<Product[]>([]);
+  private isProductInCart$: BehaviorSubject<Record<string, boolean>> = new BehaviorSubject<Record<string, boolean>> ({});
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:14]: remove this variable

File: src/app/services/product/product.service.ts [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+  constructor(private http: HttpClient) {
+    this.getProducts();
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:17]: rename to loadProduct

File: src/app/services/product/product.service.ts [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  getProducts$() {
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:20]: return type

File: src/app/services/product/product.service.ts [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+    return this.products$.asObservable();
+  }
+
+  getIsInCart() {
+    return this.isProductInCart$.asObservable();
+  }
+
+  initIsProductInCart(products: Product[]) {
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:28]: remove

File: src/app/services/product/product.service.ts [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+    let isProductInCart: Record<string, boolean> = {};
+    products.forEach( (product) => {
+      isProductInCart[product.name] = false;
+    });
+    this.updateIsInCart(isProductInCart);
+  }
+
+  getProducts() {
+    const productURL: string = 'assets/products.json';
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:37]: class property (readonly)

File: src/app/services/product/product.service.ts [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+    this.http.get<Product[]>(productURL).subscribe( products => {
+      this.updateProducts(products);
+      this.initIsProductInCart(products);
+    }, err =>  {
+      throw(err);
+    });
+  }
+
+  getProduct(productName: string): Observable<Product> {
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:46]: getProduct$()

File: src/app/services/product/product.service.ts [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+    const products: Product[] = this.products$.getValue();
+    return of(products.find(product => product.name === productName));
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:48]: use this.products$

File: src/app/services/product/product.service.ts [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  updateProducts(products: Product[]) {
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:51]: remove

File: src/app/shared/navbar/navbar.component.ts [hunk 34]
```diff
@@ -0,0 +1,34 @@
...
+export class NavbarComponent implements OnInit {
+
+  totalProducts: number = 0;
+  constructor(
+    public dialog: MatDialog,
+    private route: ActivatedRoute,
+    private cartService: CartService) { }
+
+  ngOnInit() {
+    this.cartService.getCart().subscribe(cart => {
```
Comment (alon6699) [inline src/app/shared/navbar/navbar.component.ts:21]: use totalProducts

File: src/app/cart/cart-product/cart-product.component.ts [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+     this.selectedAmount = this.amount;
+  }
+
+
+  removeFromCart(productName: string) {
+    this.remove.emit(productName);
+  }
+
+  numbersArray() {
+    return [...Array(this.product.limit).keys()].map(x => ++x);
```
[the remark this answers is not in the document]
Comment (danieTalmon) [reply src/app/cart/cart-product/cart-product.component.ts:52]: Spread syntax allows an iterable such as an array expression or string to be expanded in places where zero or more arguments (for function calls) or elements (for array literals) are expected, or an object expression to be expanded in places where zero or more key-value pairs (for object literals) are expected.

File: src/app/cart/cart.component.ts [hunk 36]
```diff
@@ -1,81 +1,66 @@
...
-    });
-    this.productService.getProducts$().subscribe(products => {
-      this.products = products;
-    });
-    this.productService.getIsInCart().subscribe(isInCart => {
-      this.isInCart = isInCart;
-    });
-  }
+    this.cartProducts$ =
+    combineLatest([this.productService.getProducts$(), this.cartService.getCart$()]).pipe(
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:25]: show me about array as argument

File: src/app/cart/cart.component.ts [hunk 37]
```diff
@@ -1,81 +1,66 @@
...
+        map(([products, cart]) => {
+          return Object.keys(cart).map(cartProductName => {
+            return {
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:28]: => { return { }}  use => ({})

File: src/app/cart/cart.component.ts [hunk 38]
```diff
@@ -1,81 +1,66 @@
...
+              product: products.find(p => p.name === cartProductName),
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:29]: rename p to product

File: src/app/cart/cart.component.ts [hunk 39]
```diff
@@ -1,81 +1,66 @@
...
   templateUrl: './cart.component.html',
   styleUrls: ['./cart.component.less']
 })
 export class CartComponent implements OnInit {
-
-  cart: Record<string, number> = {};
-  fullDetailsCart: Product[] = [];
-  products: Product[] = [];
-  isInCart: Record<string, boolean> = {};
+  cartProducts$: Observable< {product: Product, amount: number}[] >;
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:15]: try to remove spaces

File: src/app/cart/cart.component.ts [hunk 40]
```diff
@@ -1,81 +1,66 @@
...
-
-    Object.keys(this.cart).forEach(productName => {
-      this.removeFromCart(productName);
-      });
+    this.cartService.getCart$().pipe(
+      take(1)
+    ).subscribe(cart => {
+      this.productService.updateProductsLimits(cart);
+    })
+    this.cartService.checkout();
```
Comment (alon6699) [inline src/app/cart/cart.component.ts:64]: put inside

File: src/app/product-list/product-info/product-info.component.html [hunk 41]
```diff
@@ -2,10 +2,10 @@
   <div id="item-grid" class="ui one column center aligned middle aligned page grid">
     <div class="column twelve wide">
       <div class="ui centered aligned items">
-        <div class="ui centered aliened item" *ngIf="product">
+        <div class="ui centered aliened item" *ngIf="(product$ | async) as product">
             <img class= "ui medium image" src="../../assets/images/{{product.name}}.jpg">
           <div class="content">
-            <div id="product-header" class="ui large header"> {{product.name}}</div>
+            <div id="product-header" class="ui large header"> {{(product$ | async).name}}</div>
```
Comment (alon6699) [inline src/app/product-list/product-info/product-info.component.html:8]: use product

File: src/app/product-list/product-info/product-info.component.ts [hunk 42]
```diff
@@ -0,0 +1,25 @@
...
+export class ProductInfoComponent implements OnInit {
+  product$: Observable<Product>;
+
+  constructor(private productService: ProductService, private route: ActivatedRoute) { }
+
+  ngOnInit() {
+    const products$ = this.productService.getProducts$();
+    this.product$ = this.route.params.pipe(
+      switchMap((parameter: Params) => {
+        return this.productService.getProduct$(parameter.id);
```
Comment (alon6699) [inline src/app/product-list/product-info/product-info.component.ts:22]: remove {return} use arrow function => without return

File: src/app/product-list/product-list.component.ts [hunk 43]
```diff
@@ -0,0 +1,34 @@
...
+export class ProductListComponent implements OnInit {
+  products$: Observable<{product: Product, isInCart: boolean}[] >;
+
+  constructor(private productService: ProductService, private cartService: CartService) { }
+
+  ngOnInit() {
+     this.products$ = combineLatest([this.productService.getProducts$(), this.cartService.getCart$()])
+     .pipe(
+       map(([products, cart]) => {
+          return products.map(product => {
```
Comment (alon6699) [inline src/app/product-list/product-list.component.ts:21]: remove returns

File: src/app/product-list/product/product.component.html [hunk 44]
```diff
@@ -21,18 +21,23 @@
...
         </div>
         <div class="item">
           <span class="left floated">
             <p>3/5</p>
           </span>
           <span class="right floated">
-            <button (click)='isInCart? removeFromCart(product.name): addToCart(product.name)'
-            class= "ui primary button" [ngClass]="{'black':isInCart, 'primary': !isInCart, 'disabled':isOutOfStuck()}">
-            {{isInCart ? 'Remove from cart': 'Add to cart'}}
+            <button *ngIf="!isInCart; else removeButtonBlock"  (click)='addToCart()'
```
Comment (alon6699) [inline src/app/product-list/product/product.component.html:32]: without Block

File: src/app/services/product/product.service.ts [hunk 45]
```diff
@@ -1,58 +1,44 @@
...
   }
 
-  getProducts$() {
+  getProducts$(): Observable<Product[]> {
     return this.products$.asObservable();
   }
 
-  getIsInCart() {
-    return this.isProductInCart$.asObservable();
+  loadProducts() {
```
Comment (alon6699) [inline src/app/services/product/product.service.ts:22]: private

File: src/app/services/cart/cart.service.ts [hunk 46]
```diff
@@ -1,47 +1,47 @@
...
+import { map } from 'rxjs/operators';
+import { BehaviorSubject, Observable } from 'rxjs';
 
-@Injectable({
-  providedIn: 'root'
-})
+@Injectable()
 export class CartService {
-
   private cart$: BehaviorSubject<Record<string, number>> = new BehaviorSubject<Record<string, number>>({});
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:7]: init inside constructor

File: src/app/services/cart/cart.service.ts [hunk 47]
```diff
@@ -1,47 +1,47 @@
...
 
-  updateCart(cart: Record<string, number>) {
     this.cart$.next(cart);
   }
 
+  checkout() {
+    this.cart$.next({});
+  }
+
   totalProducts() {
```
Comment (alon6699) [inline src/app/services/cart/cart.service.ts:42]: return type

File: src/app/shared/navbar/navbar.component.ts [hunk 48]
```diff
@@ -10,17 +11,16 @@ import { CartComponent } from 'src/app/cart/cart.component';
   styleUrls: ['./navbar.component.less']
 })
 export class NavbarComponent implements OnInit {
+  totalProducts$: Observable<number >;
```
Comment (alon6699) [inline src/app/shared/navbar/navbar.component.ts:14]: remove space

**Pull request merged**